### PR TITLE
Worksheet Generator Feature PR

### PR DESCRIPTION
**Pull request by vash02**
# Worksheet Question Generator

This project enables the generation of quiz and worksheet questions using a machine learning model, integrated with LangChain. Users can provide topics, difficulty levels, and hints to generate customized questions, with robust validation processes to ensure accuracy and relevance.

### Dual-Prompt Setup
Based on experimentation, using two distinct prompts—one for quiz questions and another for worksheet questions—yields optimal results. This dual-prompt approach enhances the quality and relevance of the generated content.

### Question Generation
The `WorksheetBuilder` class supports the generation of both quiz and worksheet questions based on:
- **Topic**: Subject of the questions.
- **Level**: Difficulty (beginner, intermediate, advanced).
- **Hint**: Question style (e.g., "single sentence" or "multiple choice").
- **q_type**: Specifies quiz or worksheet.

### Validation Mechanism
- **Format Validation**: Ensures quiz questions contain choices, answers, and explanations; worksheet questions contain a question, answer, and explanation.
- **Cosine Similarity Validation**: Ensures question-answer relevance using cosine similarity scores, rejecting hallucinations with a similarity threshold.

### Logging & Error Handling
Detailed logs at key stages ensure smooth execution and help debug any issues.

## How to Use
1. Instantiate the `WorksheetBuilder` class with the required parameters.
2. Use `create_questions()` for quiz questions or `create_worksheet_questions()` for worksheet questions.
3. The validated questions are returned as dictionaries containing the question, answer, and explanation.

### Example
```python
executor([ToolFile(url="https://example.com/resource.pdf")],
         "machine learning", "Masters", "single sentence answer questions", 5, 5)

### CODE REVIEW

File: app/utils/auth.py [hunk 1]
```diff
@@ -14,11 +14,13 @@ def access_secret_file(secret_id, version_id="latest"):
 
 # Function to ensure incoming request is from controller with key
 def key_check(api_key: str = Header(None)):
-  
+  print("api key",api_key)
```
Comment (hash2004) [inline app/utils/auth.py:17]: Do not print the environment variables anywhere. It is unsafe and is not good practice.

File: app/utils/auth.py [hunk 2]
```diff
@@ -14,11 +14,13 @@ def access_secret_file(secret_id, version_id="latest"):
...
+  print("env type",os.environ['ENV_TYPE'])
```
Comment (hash2004) [inline app/utils/auth.py:18]: Using print statements to display environment variables, such as `print("env type", os.environ['ENV_TYPE'])`, is not considered a best practice. Printing sensitive information like environment types can lead to security risks and clutter the output with unnecessary details. Instead, consider using proper logging frameworks or avoid printing such values unless absolutely necessary.

**Review by hash2004 — CHANGES_REQUESTED** [review]
Please add description to your PR as well.  You can use this [pull request as inspiration](https://github.com/radicalxdev/marvel-ai-backend/pull/101).

File: app/features/worksheet_generator/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+## How to Use
+
+1. Instantiate the `WorksheetBuilder` class with the necessary parameters, including the topic, level, hint, and question type.
+2. Use one of the two dedicated prompts based on your requirements:
+   - For quiz generation, call the `create_questions()` method.
+   - For worksheet generation, use the `create_worksheet_questions()` method.
+3. The system will generate, validate, and return the questions as a list of dictionaries. Each dictionary contains a question, its answer, and an explanation.
+
+## Example Usage
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/README.md:51]: Good README.md. You can try adding the request and response interfaces for enabling a better understanding of what the dev team can expect to interact with during the requests to the AI endpoint.

File: app/features/worksheet_generator/core.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+# import os
+from features.quizzify.tools import RAGpipeline
+from services.tool_registry import ToolFile
+from services.logger import setup_logger
+from features.worksheet_generator.tools import WorksheetBuilder
+from api.error_utilities import LoaderError, ToolExecutorError
+logger = setup_logger()
+
+
+def executor(files: list[ToolFile], topic: str, level: str, hint: str, hint_num: int, num_questions: int, verbose=True):
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/core.py:11]: You can use a Pydantic schema for decoupling the args.

File: app/features/worksheet_generator/core.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    try:
+        if verbose: logger.debug(f"Files: {files}")
+        # Instantiate RAG pipeline with default values
+        pipeline = RAGpipeline(verbose=verbose)
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/core.py:15]: Chage the name of this class as it does not support a complete understanding of what it does for the Worksheet Generator.

File: app/features/worksheet_generator/core.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        pipeline.compile()
+        # Process the uploaded files
+        db = pipeline(files)
+
+        # Create and return the quiz questions
+        output = WorksheetBuilder(db, topic, level, hint, "quiz").create_questions(num_questions)
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/core.py:21]: You can instantiate your WorksheetBuilder class first instead for reusing the same variable and improve the readability.

File: app/features/worksheet_generator/prompt/worksheet-generator-worksheet-prompt.txt
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/prompt/worksheet-generator-worksheet-prompt.txt:1]: What's the difference between this file and the other `worksheet-generator-quiz-prompt.txt`? If they are different, please make sure to use different names for better readability.

File: app/features/worksheet_generator/thebook.pdf
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/thebook.pdf:1]: Why is this file required?

File: app/features/worksheet_generator/tools.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+    # Combine the script directory with the relative file path
+    absolute_file_path = os.path.join(script_dir, file_path)
+
+    with open(absolute_file_path, 'r') as file:
+        return file.read()
+class WorksheetBuilder:
+    def __init__(self, db, topic, level=None, hint=None, q_type=""):
+        default_config = {
+            "model": VertexAI(model="gemini-1.0-pro"),
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:31]: VertexAI is not used anymore in the main project.

File: app/features/worksheet_generator/tools.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    # Get the directory containing the script file
+    script_dir = os.path.dirname(os.path.abspath(__file__))
+
+    # Combine the script directory with the relative file path
+    absolute_file_path = os.path.join(script_dir, file_path)
+
+    with open(absolute_file_path, 'r') as file:
+        return file.read()
+class WorksheetBuilder:
+    def __init__(self, db, topic, level=None, hint=None, q_type=""):
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:29]: You can create a Pydantic schema for decoupling the args.

File: app/features/worksheet_generator/tools.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+        # if self.q_type == "quiz":
+        prompt = self.generate_prompt(self.topic, self.level, self.hint, self.parser[self.q_type], self.q_type)
+        #
+        # if self.q_type == "worksheet":
+        #     prompt = self.generate_prompt(self.topic, self.level, self.hint, self.parser[self.q_type], self.q_type)
+
+        retriever = self.db.as_retriever()
+
+        runner = RunnableParallel({"context": retriever, "topic":RunnablePassthrough(), "hint": RunnablePassthrough(),
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:71]: Good parallelization of the AI pipeline.

File: app/features/worksheet_generator/tools.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        retrieval_chain_quiz = (runner | prompt | self.model | self.parser["quiz"])
+
+        retrieval_chain_worksheet = (runner | prompt | self.model | self.parser["worksheet"])
+
+        if self.verbose:
+            print("Chain successfully created",retrieval_chain_quiz,retrieval_chain_worksheet)
+        if self.q_type == "quiz": return retrieval_chain_quiz
+        if self.q_type == "worksheet": return retrieval_chain_worksheet
+
+    def validate_response(self, response: dict) -> bool:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:85]: This is not required as the JSONOutputParser supports this.

File: app/features/worksheet_generator/tools.py [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+                            if not isinstance(key, str) or not isinstance(value, str):
+                                return False
+                        return True
+            return False
+        except TypeError as e:
+            if self.verbose:
+                logger.error(f"TypeError during response validation: {e}")
+            return False
+
+    def validate_worksheet_response(self, response: dict) -> bool:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:102]: This is not required as the JSONOutputParser supports this.

File: app/features/worksheet_generator/tools.py [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+            if isinstance(response, dict):
+                if 'question' in response and 'answer' in response and 'explanation' in response:
+                        return True
+            return False
+        except TypeError as e:
+            if self.verbose:
+                logger.error(f"TypeError during response validation: {e}")
+            return False
+
+    def is_valid_dict_values(self, json_dict):
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:114]: This is not required as the JSONOutputParser supports this.

File: app/features/worksheet_generator/tools.py [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        """
+        Check if the dictionary values contain special or newline characters.
+        """
+        for key, value in json_dict.items():
+            if isinstance(value, str) and re.search(r'[\n\r]', value):
+                return False
+        return True
+
+    def format_choices(self, choices: Dict[str, str]) -> List[Dict[str, str]]:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:123]: This is not required as the JSONOutputParser supports this.

File: app/features/worksheet_generator/tools.py [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        if num_questions > 10:
+            return {"message": "error", "data": "Number of questions cannot exceed 10"}
+
+        chain = self.execute()
+
+        generated_questions = []
+        attempts = 0
+        max_attempts = num_questions * 5  # Allow for more attempts to generate questions
+        print("len of gen qs", len(generated_questions))
+        while len(generated_questions) < num_questions and attempts < max_attempts:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:138]: Ensure that this while loop is not costful for optimal results.

File: app/features/worksheet_generator/tools.py [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        if len(generated_questions) < num_questions:
+            logger.warning(f"Only generated {len(generated_questions)} out of {num_questions} requested questions")
+
+        if len(generated_questions) == 10:
+            if self.verbose: logger.info(f"10 questions generated, Deleting vectorstore")
+            self.db.delete_collection()
+
+        # Return the list of questions
+        return generated_questions[:num_questions]
+    def validate_qtype_response(self, hint: str, answer) -> bool:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:167]: This is not required as the JSONOutputParser supports this.

File: app/features/worksheet_generator/tools.py [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+        if "single sentence" in hint:
+            return 10 <= len(answer.split()) <= 400
+        if "integer" in hint:
+            try:
+                int(answer)
+                return True
+            except ValueError:
+                return False
+
+    def is_response_relevant(self, response: dict) -> bool:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:177]: This proposal is interesting but costful in terms of production resources. Indeed, Google actually provides metrics to accomplish this semantic search process with their embeddings.

File: app/features/worksheet_generator/tools.py [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+
+    def create_worksheet_questions(self, hint_num=3):
+        chain = self.execute()
+
+        generated_questions = []
+        attempts = 0
+        max_attempts = hint_num * 5
+
+        print("len of gen qs", len(generated_questions))
+        while len(generated_questions) < hint_num and attempts < max_attempts:
```
Comment (AaronSosaRamos) [inline app/features/worksheet_generator/tools.py:212]: Ensure that this while loop is not costful in terms of token generation.

File: app/utils/auth.py [hunk 18]
```diff
@@ -14,11 +14,11 @@ def access_secret_file(secret_id, version_id="latest"):
 
 # Function to ensure incoming request is from controller with key
 def key_check(api_key: str = Header(None)):
-  
   if os.environ['ENV_TYPE'] == "production":
     set_key = access_secret_file("backend-access")
   else:
     set_key = "dev"
-  
+  if api_key is None:
```
Comment (AaronSosaRamos) [inline app/utils/auth.py:21]: Please, remove this as it is affecting the auth. of the AI endpoint.

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Please, solve the given comments

**Comment by AaronSosaRamos** [comment]
Closed due inactivity

**Pull request closed without merging**